File: scripts/release.mjs
#!/usr/bin/env node

/**
 * Release script for agent-enhance-kit
 * 
 * This script:
 * 1. Reads version from packages/aek-websearch/VERSION
 * 2. Checks if tag already exists
 * 3. Creates git tag
 * 4. Pushes tag to GitHub
 * 
 * Used as a "finally" script in cqg acp configuration.
 * Config path: .cheezmil_quick_git/config/which_script_run_when_quick_add_commit_push_execute_finally.txt
 */

import { execSync } from 'node:child_process';
import { readFileSync, existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Paths
const ROOT_DIR = join(__dirname, '..');
const VERSION_FILE = join(ROOT_DIR, 'packages', 'aek-websearch', 'VERSION');

// Colors
const colors = {
    reset: '\x1b[0m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    cyan: '\x1b[36m',
};

function log(message, color = 'reset') {
    console.log(`${colors[color] || colors.reset}${message}${colors.reset}`);
}

function execCommand(command, options = {}) {
    try {
        return execSync(command, {
            cwd: ROOT_DIR,
            encoding: 'utf-8',
            stdio: ['pipe', 'pipe', 'pipe'],
            ...options,
        }).trim();
    } catch (error) {
        return null;
    }
}

function main() {
    log('\n=== Agent Enhance Kit Release ===\n', 'cyan');

    // 1. Check if VERSION file exists
    if (!existsSync(VERSION_FILE)) {
        log(`ERROR: VERSION file not found at ${VERSION_FILE}`, 'red');
        process.exit(1);
    }

    // 2. Read version
    const version = readFileSync(VERSION_FILE, 'utf-8').trim();
    if (!version) {
        log('ERROR: VERSION file is empty', 'red');
        process.exit(1);
    }

    const tagName = `v${version}`;
    log(`Version: ${version}`, 'green');
    log(`Tag: ${tagName}`, 'green');

    // 3. Check if tag already exists
    const existingTag = execCommand(`git tag -l "${tagName}"`);
    if (existingTag === tagName) {
        log(`\nTag ${tagName} already exists. Skipping release.`, 'yellow');
        log('If you want to re-release, delete the tag first:', 'yellow');
        log(`  git tag -d ${tagName}`, 'yellow');
        log(`  git push origin :refs/tags/${tagName}`, 'yellow');
        return;
    }

    // 4. Check if there are uncommitted changes
    const status = execCommand('git status --porcelain');
    if (status) {
        log('\nWARNING: There are uncommitted changes.', 'yellow');
        log('Please commit or stash changes before release.', 'yellow');
        process.exit(1);
    }

    // 5. Create tag
    log(`\nCreating tag ${tagName}...`, 'cyan');
    const tagResult = execCommand(`git tag -a ${tagName} -m "Release ${tagName}"`);
    if (tagResult === null) {
        log('ERROR: Failed to create tag', 'red');
        process.exit(1);
    }
    log(`Tag ${tagName} created successfully`, 'green');

    // 6. Push tag to GitHub
    log(`\nPushing tag ${tagName} to GitHub...`, 'cyan');
    const pushResult = execCommand(`git push origin ${tagName}`);
    if (pushResult === null) {
        log('ERROR: Failed to push tag', 'red');
        log('You may need to push manually:', 'yellow');
        log(`  git push origin ${tagName}`, 'yellow');
        process.exit(1);
    }

    log(`\n✅ Release ${tagName} initiated successfully!`, 'green');
    log('GitHub Actions will now build and publish the release.', 'cyan');
    log(`Check progress at: https://github.com/cheezmil/agent-enhance-kit/actions`, 'cyan');
}

main();

File: scripts/for-maintainers/release.py
#!/usr/bin/env python3

import json
import sys
import subprocess
from pathlib import Path

# Paths
SCRIPT_DIR = Path(__file__).parent
ROOT_DIR = SCRIPT_DIR.parent.parent  # up to scripts, then to root
VERSION_FILE = ROOT_DIR / "VERSION"
PACKAGE_JSON = ROOT_DIR / "package.json"

# Colors
class Colors:
    RESET = '\033[0m'
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    CYAN = '\033[36m'

def log(message, color=Colors.RESET):
    print(f"{color}{message}{Colors.RESET}")

def exec_command(command, check=False):
    """Execute a shell command and return output."""
    try:
        result = subprocess.run(
            command,
            cwd=ROOT_DIR,
            shell=True,
            capture_output=True,
            text=True,
            check=check
        )
        return result.stdout.strip()
    except subprocess.CalledProcessError:
        return None

def get_version():
    """Read version from root package.json."""
    if not PACKAGE_JSON.exists():
        return None
    
    try:
        data = json.loads(PACKAGE_JSON.read_text())
        return data.get("version")
    except json.JSONDecodeError:
        return None

def main():
    print()
    log("=== Agent Enhance Kit Release ===", Colors.CYAN)
    print()

    # 1. Get version from package.json
    version = get_version()
    if not version:
        log("ERROR: Version not found in package.json", Colors.RED)
        sys.exit(1)

    log(f"Version: {version}", Colors.GREEN)

    # 3. Check if there are uncommitted changes
    status = exec_command("git status --porcelain")
    if status:
        print()
        log("WARNING: There are uncommitted changes.", Colors.YELLOW)
        log("Please commit or stash changes before release.", Colors.YELLOW)
        log("Run: pnpm changeset && git add . && git commit && git push", Colors.YELLOW)
        sys.exit(1)

    # 4. Check if changesets are pending
    changeset_status = exec_command("pnpm changeset status")
    if "No unreleased changesets found" not in (changeset_status or ""):
        print()
        log("There are pending changesets. Please run 'pnpm changeset' first.", Colors.YELLOW)
        sys.exit(1)

    # 5. Create tags for each package
    log(f"Creating tags for v{version}...", Colors.CYAN)
    
    packages = ["aek-websearch", "aek-task-manager"]
    tags_created = []
    
    for pkg in packages:
        tag_name = f"{pkg}@v{version}"
        existing_tag = exec_command(f'git tag -l "{tag_name}"')
        if existing_tag == tag_name:
            log(f"Tag {tag_name} already exists. Skipping.", Colors.YELLOW)
            continue
        
        tag_result = exec_command(f'git tag -a {tag_name} -m "Release {tag_name}"')
        if tag_result is None:
            log(f"ERROR: Failed to create tag {tag_name}", Colors.RED)
            continue
        
        log(f"Tag {tag_name} created", Colors.GREEN)
        tags_created.append(tag_name)

    if not tags_created:
        log("No new tags to create.", Colors.YELLOW)
        return

    # 6. Push tags
    print()
    log("Pushing tags to GitHub...", Colors.CYAN)
    push_result = exec_command("git push github --tags")
    if push_result is None:
        log("ERROR: Failed to push tags", Colors.RED)
        log("You may need to push manually:", Colors.YELLOW)
        for tag in tags_created:
            log(f"  git push github {tag}", Colors.YELLOW)
        sys.exit(1)

    # 7. Trigger GitHub Actions for npm publish
    print()
    log(f"✅ Release v{version} initiated!", Colors.GREEN)
    log("GitHub Actions will now build and publish to npm.", Colors.CYAN)
    log("Check progress at: https://github.com/cheezmil/agent-enhance-kit/actions", Colors.CYAN)

if __name__ == "__main__":
    main()
